agent/pkg/singularity/singularity.go

```diff
@@ -387,6 +378,49 @@ func (s *SingularityClient) RunContainer(
 	}, nil
 }
 
+// Sets the environment of the process that will run the Singularity/apptainer command.
+func (s *SingularityClient) setCommandEnvironment(
+	req cproto.RunSpec, cudaVisibleDevices []string, cmd *exec.Cmd,
+) {
+	// Per https://pkg.go.dev/os/exec#Cmd.Env, if cmd.Env is nil, the new process uses the current
+	// process's environment. If this in not the case, for example because we specify something to
+	// control Singularity operation, then we need to specify everything needed from the current
+	// environment,
+	if req.DeviceType == device.CUDA {
+		cudaVisibleDevicesVar := strings.Join(cudaVisibleDevices, ",")
+		cmd.Env = append(cmd.Env,
+			fmt.Sprintf("SINGULARITYENV_CUDA_VISIBLE_DEVICES=%s", cudaVisibleDevicesVar),
+			fmt.Sprintf("APPTAINERENV_CUDA_VISIBLE_DEVICES=%s", cudaVisibleDevicesVar),
+		)
+	}
+	if req.DeviceType == device.ROCM {
+		// Avoid this problem: https://github.com/determined-ai/determined-ee/pull/922
+		// by not setting both ROCR_VISIBLE_DEVICES & CUDA_VISIBLE_DEVICES
+		s.addToCommandEnvironment(cmd, "ROCR_VISIBLE_DEVICES")
+	}
+
+	s.addOptionalRegistryAuthCredentials(req, cmd)
+
+	// HACK(singularity): without this, --nv doesn't work right. If the singularity run command
+	// cannot find nvidia-smi, the --nv fails to make it available inside the container, e.g.,
+	// env -i /usr/bin/singularity run --nv \\
+	//   docker://determinedai/environments:cuda-11.3-pytorch-1.10-tf-2.8-gpu-24586f0 nvidia-smi
+	cmd.Env = append(cmd.Env, fmt.Sprintf("PATH=%s", os.Getenv("PATH")))
+
+	s.addToCommandEnvironment(cmd, "http_proxy")
+	s.addToCommandEnvironment(cmd, "https_proxy")
+	s.addToCommandEnvironment(cmd, "no_proxy")
+	s.addToCommandEnvironment(cmd, "APPTAINER_CACHEDIR")
+	s.addToCommandEnvironment(cmd, "SINGULARITY_CACHEDIR")
+}
+
+func (s *SingularityClient) addToCommandEnvironment(cmd *exec.Cmd, variable string) {
```

Comment:
The prior name include ifSet suffix -- it still only adds to the environment if set, so seems better to include that in the name.

Reply:
OK